src/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs

```diff
@@ -0,0 +1,92 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Http
+{
+    /// <summary>
+    /// DiagnosticHandler notifies DiagnosticSource subscribers about outgoing Http requests
+    /// </summary>
+    internal sealed class DiagnosticsHandler : DelegatingHandler
+    {
+        /// <summary>
+        /// DiagnosticHandler constructor
+        /// </summary>
+        /// <param name="innerHandler">Inner handler: Windows or Unix implementation of HttpMessageHandler. 
+        /// Note that DiagnosticHandler is the latest in the pipeline </param>
+        public DiagnosticsHandler(HttpMessageHandler innerHandler) : base(innerHandler)
+        {
+        }
+
+        internal static bool IsEnabled()
+        {
+            return s_diagnosticListener.IsEnabled(DiagnosticsHandlerLoggingStrings.ResponseWriteName);
+        }
+
+        protected internal override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request,
+            CancellationToken cancellationToken)
+        {
+            //HttpClientHandler is responsible to call DelegatingHandler.IsEnabled() before forwarding request here.
+            //This code will not be reached if ResponseWriteName is not enabled, unless consumer unsubscribes
+            //from DiagnosticListener right after the check. So some requests happening right after subscription starts
+            //might not be instrumented. Similarly, when consumer unsubscribes, extra requests might be instumented
+
+            Guid loggingRequestId = Guid.NewGuid();
+            if (s_diagnosticListener.IsEnabled(DiagnosticsHandlerLoggingStrings.RequestWriteName))
+            {
+                LogHttpRequest(request,  loggingRequestId);
+            }
+            Task<HttpResponseMessage> responseTask = base.SendAsync(request, cancellationToken);
+
+            LogHttpResponse(responseTask, loggingRequestId);
+            return responseTask;
```

Comment:
Why pass in the task, add a continuation, and then return the same task, instead of just using the async-await feature like "return LogHttpResponse(await base.SendAsync(request, cancellationToken), loggingRequestId)"?